src/core/algorithms/nd/nd_verifier/util/stats_calculator.cpp

```diff
@@ -0,0 +1,118 @@
+#include "algorithms/nd/nd_verifier/util/stats_calculator.h"
+
+#include <algorithm>
+#include <chrono>
+#include <memory>
+#include <numeric>
+#include <string>
+#include <vector>
+
+#include <easylogging++.h>
+
+namespace algos::nd_verifier::util {
+
+template <typename It>
+std::shared_ptr<std::vector<size_t>> StatsCalculator::CalculateFrequencies(size_t codes_number,
+                                                                           It begin, It end) {
+    auto result = std::make_shared<std::vector<size_t>>();
+    for (size_t code{0}; code < codes_number; ++code) {
+        result->push_back(std::count(begin, end, code));
+    }
+
+    return result;
+}
+
+void StatsCalculator::CalculateStats() {
+    auto start_time = std::chrono::system_clock::now();
+    lhs_frequencies_ =
+            CalculateFrequencies(lhs_values_->size(), encoded_lhs_->begin(), encoded_lhs_->end());
+    rhs_frequencies_ =
+            CalculateFrequencies(rhs_values_->size(), encoded_rhs_->begin(), encoded_rhs_->end());
+
+    std::vector<size_t> highlights_lhs_codes;
+    model::WeightType max_weight{0};
+    model::WeightType min_weight{UINT_MAX};
+    for (size_t code{0}; code < lhs_values_->size(); ++code) {
+        auto weight = value_deps_->at(code).size();
+
+        assert(weight > 0);
```

Comment:
`assert` macro makes use only for debug build. I think that it is better to use exceptions here if necessary.